### PR TITLE
add cutout image op

### PR DESCRIPTION
**Pull request by fsx950223**
Related #1333

### CODE REVIEW

File: tensorflow_addons/image/cutout_ops_test.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        test_image_file = tf.io.read_file(
+            "tensorflow_addons/image/test_data/Yellow_Smiley_Face.png"
+        )
+        expect_image_file = tf.io.read_file(
+            "tensorflow_addons/image/test_data/Yellow_Smiley_Face_Warp-cutout-20.png"
+        )
+        for channel in [0, 1, 3, 4]:
+            with self.subTest(channel=channel):
+                test_image = tf.image.decode_image(
+                    test_image_file, channels=channel, dtype=tf.uint8
```
[the remark this answers is not in the document]
Comment (fsx950223) [reply tensorflow_addons/image/cutout_ops_test.py]: I have tested ```dtype``` in ```test_different_dtypes``` and it will increase test time if I use it on all test cases.

File: tensorflow_addons/image/cutout_ops.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            tf.zeros(cutout_shape, dtype=image.dtype), padding_dims, constant_values=1
+        )
+
+        if tf.equal(tf.rank(image), 3):
+            mask = tf.expand_dims(mask, -1)
+            mask = tf.tile(mask, [1, 1, tf.shape(image)[-1]])
+        elif tf.equal(tf.rank(image), 4):
+            mask = tf.expand_dims(mask, 0)
+            mask = tf.expand_dims(mask, -1)
+            mask = tf.tile(mask, [tf.shape(image)[0], 1, 1, tf.shape(image)[-1]])
```
[the remark this answers is not in the document]
Comment (fsx950223) [reply tensorflow_addons/image/cutout_ops.py]: Changed. But I wonder about the performance of these ops.

**Comment by gabrieldemarmiesse** [comment]
Thank you for this pull request! I believe we're going to change the API of the reference implementation to make it more user friendly. Here is what I propose:

```python

def cutout(
    images: TensorLike,
    mask_size: TensorLike,
    offset: TensorLike= (0, 0),
    constant_values: Number = 0,
    data_format="channels_last"
) -> tf.Tensor:
"""

Args:
    image: A 4D tensor, (N, H, W, C). 
    mask_size: a tuple (or tensor) with two values. Height and width of the mask (correspond to 2x pad_size). A single scalar means a square mask (like in keras conv and pad layers). Can be a tensor of shape (N, 2) to have different mask sizes in the same batch.
    offset: the offset relative to the center of the image. Default is (0, 0) which means the mask will be in the middle of the image. Can be a tensor of shape (N, 2) to have different offsets in the same batch.
    constant_values: the values used to fill the mask.
    data_format: The data format
"""
```

The rational behind using only 4D tensors is that when the users passes 3D tensors, there is a non-negligable probability that user mistake `NHW` and `HWC`.

It's totally fine to not implement everything. For example, you can do 
```python
if data_format == "channels_first":
    raise NotImplementedError("Channels first is not yet available for cutout. Contributions welcome!")

if tf.rank(mask_size) != 0:
     raise NotImplementedError("Having non-square masks is not supported yet, contributions welcome.")
```


random_cutout should follow a similar pattern.

I believe that it should not change too much your code, it's already really close to the end result.

For the tests, please avoid adding images to the git repo, also we moving away from `tf.test.TestCase` and `@run_all_in_graph_and_eager_mode`. The philosophy now is to use plain pytest tests with the decorator `@pytest.mark.usefixtures("maybe_run_functions_eagerly")` . This enables eager mode in tests which helps a lot in readability. See #1328 .

**Comment by gabrieldemarmiesse** [comment]
@fsx950223 feel free to ping me when you're done with the changes.

**Comment by fsx950223** [comment]
cc @gabrieldemarmiesse

File: tensorflow_addons/image/cutout_ops.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    """
+    mask_size, data_format, image_height, image_width = _norm_params(
+        images, mask_size, data_format
+    )
+
+    cutout_center_height = tf.random.uniform(
+        shape=[], minval=0, maxval=image_height, dtype=tf.int32, seed=seed
+    )
+    cutout_center_width = tf.random.uniform(
+        shape=[], minval=0, maxval=image_width, dtype=tf.int32, seed=seed
```
Comment (gabrieldemarmiesse) [inline tensorflow_addons/image/cutout_ops.py]: If I understand correctly, if a batch of images is passed and random cutout is applied, the mask will be at the same place for all the images in the batch, right? Is this something we want? Maybe users would expect (as an augmentation strategy) to have the mask being at different places for all the images in the batch. 
If the original implementation uses the same mask for all the images in the batch, let's go with this implementation, but the function shouldn't be public. I believe a public `random_cutout` shoud respect the principle of least astonishment. 
What do you think?

**Comment by fsx950223** [comment]
> Thank you for the pull request. We're nearly there, a few changes here and there need to be made.

I will change them later.

**Comment by fsx950223** [comment]
cc @gabrieldemarmiesse

File: tensorflow_addons/image/cutout_ops.py (hunk 3 above)
Comment (AakashKumarNain) [reply tensorflow_addons/image/cutout_ops.py]: Exactly! In a batch the cutout operation has to be random for each image.

File: tensorflow_addons/image/cutout_ops.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    value `replace`. The located where the mask will be applied is randomly
+    chosen uniformly over the whole images.
+
+    Args:
+      images: A tensor of shape
+        (batch_size, height, width, channels)
+        (NHWC), (batch_size, channels, height, width)(NCHW).
+      mask_size: Specifies how big the zero mask that will be generated is that
+        is applied to the images. The mask will be of size
+        (2 * mask_height x 2 * mask_width).
```
Comment (gabrieldemarmiesse) [inline tensorflow_addons/image/cutout_ops.py]: `mask_size` should be the shape of the mask. Here the shape of the true mask is `(2 * mask_height x 2 * mask_width)` which might confuse users. Could you change that to `(mask_height x mask_width)`? You can throw an error if the width and height cannot be divided by 2.

File: tensorflow_addons/image/cutout_ops.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    a random location within `img`. The pixel values filled in will be of the
+    value `replace`. The located where the mask will be applied is randomly
+    chosen uniformly over the whole images.
+
+    Args:
+      images: A tensor of shape (batch_size, height, width, channels)
+        (NHWC), (batch_size, channels, height, width)(NCHW).
+      mask_size: Specifies how big the zero mask that will be generated is that
+        is applied to the images. The mask will be of size
+        (2 * mask_height x 2 * mask_width).
```
Comment (gabrieldemarmiesse) [inline tensorflow_addons/image/cutout_ops.py]: Same as above.

File: tensorflow_addons/image/cutout_ops.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+      offset: A tuple of (height, width)
```
Comment (gabrieldemarmiesse) [inline tensorflow_addons/image/cutout_ops.py]: Could you specify all the shapes are are possible? A tuple is possible but a 2D tensor too (batch_size, 2)

File: tensorflow_addons/image/cutout_ops.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    images: TensorLike,
+    mask_size: TensorLike,
+    offset: TensorLike = (0, 0),
+    constant_values: Number = 0,
+    data_format: str = "channels_last",
+) -> tf.Tensor:
+    """Apply cutout (https://arxiv.org/abs/1708.04552) to images.
+
+    This operation applies a (2 * mask_height x 2 * mask_width) mask of zeros to
+    a random location within `img`. The pixel values filled in will be of the
```
Comment (gabrieldemarmiesse) [inline tensorflow_addons/image/cutout_ops.py]:
```suggestion
    a location within `img` specified by the offset. The pixel values filled in will be of the
```

File: tensorflow_addons/image/cutout_ops_test.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    result_image = random_cutout(
+        test_image, 20, seed=1234, data_format="channels_first"
+    )
+    np.testing.assert_allclose(tf.shape(result_image), tf.shape(expect_image))
+
+
+@pytest.mark.usefixtures("maybe_run_functions_eagerly")
+def test_with_tf_function():
+    test_image = tf.ones([1, 40, 40, 1], dtype=tf.uint8)
+    result_image = tf.function(random_cutout)(test_image, 2)
```
Comment (gabrieldemarmiesse) [inline tensorflow_addons/image/cutout_ops_test.py]: Here we need to make sure that the for loop with the TensorArray works well. To do that, we need to make sure the loop is possible even if the size of the batch is not known when drawing the graph. Could you use `input_signature` and set the shape of `images` to `[None, 40, 40, 1]` ?

File: tensorflow_addons/image/cutout_ops_test.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    cutout_area = tf.zeros([4, 4], dtype=tf.uint8)
+    cutout_area = tf.pad(cutout_area, ((0, 36), (0, 36)), constant_values=1)
+    expect_image = to_4D_image(cutout_area)
+    np.testing.assert_allclose(tf.shape(result_image), tf.shape(expect_image))
+
+
+if __name__ == "__main__":
```
Comment (gabrieldemarmiesse) [inline tensorflow_addons/image/cutout_ops_test.py]: Could you add one more test to ensure that with random cutout, the masks are at different places in the images of the batch? I'll let you decide what's the easier way of doing that.

File: tensorflow_addons/image/cutout_ops_test.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+import tensorflow as tf
+import numpy as np
+from absl.testing import parameterized
+from tensorflow_addons.image.cutout_ops import cutout, random_cutout
+from tensorflow_addons.image.utils import to_4D_image
+
+
+@parameterized.named_parameters(
+    ("float16", np.float16), ("float32", np.float32), ("uint8", np.uint8)
+)
```
Comment (gabrieldemarmiesse) [inline tensorflow_addons/image/cutout_ops_test.py]:
```suggestion
@pytest.mark.parameterize("dtype", [np.float16, np.float32, np.uint8])
```

File: tensorflow_addons/image/cutout_ops_test.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+def test_different_dtypes(dtype):
+    test_image = tf.ones([1, 40, 40, 1], dtype=dtype)
+    result_image = cutout(test_image, 2, [2, 2])
+    cutout_area = tf.zeros([4, 4], dtype=dtype)
+    cutout_area = tf.pad(cutout_area, ((0, 36), (0, 36)), constant_values=1)
+    expect_image = to_4D_image(cutout_area)
+    np.testing.assert_allclose(result_image, expect_image)
+
+
```
Comment (gabrieldemarmiesse) [inline tensorflow_addons/image/cutout_ops_test.py]:
```suggestion
    np.testing.assert_allclose(result_image, expect_image)
    assert result_image.dtype == dtype


```

**Comment by fsx950223** [comment]
cc @gabrieldemarmiesse .

File: tensorflow_addons/image/cutout_ops.py [hunk 12]
```diff
@@ -30,11 +30,20 @@ def _get_image_wh(images, data_format):
...
-    return mask_size, data_format, image_height, image_width
+    with tf.control_dependencies(
+        [
+            tf.assert_equal(
+                tf.reduce_any(mask_size % 2 != 0),
+                False,
+                "mask_size should be divisible by 2",
+            )
+        ]
+    ):
```
Comment (gabrieldemarmiesse) [inline tensorflow_addons/image/cutout_ops.py]: control_dependencies is quite tf 1.x style, I recommend to activate the check only in eager mode. It will also be easier to debug for users.

Comment (gabrieldemarmiesse) [reply tensorflow_addons/image/cutout_ops.py]: Your piece of code raised a very important question that I discussed in #1417 . We don't need to fix #1417 for merging your pull request of course.

**Comment by fsx950223** [comment]
> Thanks for this great pull request!

Thanks for your review
